File: src/cpp/config/LlmConfig.cpp
//
// SPDX-FileCopyrightText: Copyright 2024-2025 Arm Limited and/or its affiliates <open-source-office@arm.com>
//
// SPDX-License-Identifier: Apache-2.0
//
#include "LlmConfig.hpp"
#include <stdexcept>

LlmConfig::LlmConfig(const std::string& modelTag,
                     const std::string& userTag,
                     const std::string& endTag,
                     const std::string& modelPath,
                     const std::string& llmPrefix,
                     int numThreads,
                     int batchSize) :
    m_modelTag(modelTag), m_userTag(userTag), m_endTag(endTag), m_modelPath(modelPath), m_llmPrefix(llmPrefix)
{
    SetNumThreads(numThreads);
    SetBatchSize(batchSize);
}

std::string LlmConfig::GetEndTag() const
{
    return this->m_endTag;
}

std::string LlmConfig::GetUserTag() const
{
    return this->m_userTag;
}

std::string LlmConfig::GetModelTag() const
{
    return this->m_modelTag;
}

std::string LlmConfig::GetModelPath() const
{
    return this->m_modelPath;
}

std::string LlmConfig::GetLlmPrefix() const
{
    return this->m_llmPrefix;
}

int LlmConfig::GetNumThreads() const
{
    return this->m_numThreads;
}

int LlmConfig::GetBatchSize() const
{
    return this->m_batchSize;
}

void LlmConfig::SetModelTag(const std::string& modelIdentifier)
{
    this->m_modelTag = modelIdentifier;
}

void LlmConfig::SetUserTag(const std::string& userTag)
{
    this->m_userTag = userTag;
}

void LlmConfig::SetEndTag(const std::string& endTag)
{
    this->m_endTag = endTag;
}

void LlmConfig::SetModelPath(const std::string& basePath)
{
    this->m_modelPath = basePath;
}

void LlmConfig::SetLlmPrefix(const std::string& llmInitialPrompt)
{
    this->m_llmPrefix = llmInitialPrompt;
}

void LlmConfig::SetNumThreads(int threads)
{
    if (threads <= 0) {
        throw std::invalid_argument("number of threads must be a positive integer.");
    }
    this->m_numThreads = threads;
}

void LlmConfig::SetBatchSize(int batchSz)
{
    if (batchSz <= 0) {
        throw std::invalid_argument("batch-size must be a positive integer.");
    }
    this->m_batchSize = batchSz;
}

File: src/java/com/arm/LlmConfig.java
//
// SPDX-FileCopyrightText: Copyright 2025 Arm Limited and/or its affiliates <open-source-office@arm.com>
//
// SPDX-License-Identifier: Apache-2.0
//

package com.arm;

import java.util.List;

/**
 * LlmConfig class for adding the the settings for the Large Language Model.
 */
public class LlmConfig
{
    private String modelTag;
    private String userTag;
    private String endTag;
    private String modelPath;
    private String llmPrefix;
    private List<String> stopWords;
    private int numThreads;

    /**
     * Minimal constructor without userTag, endTag and numThreads
     *
     * @param modelTag   tag for the model
     * @param stopWords  stop words to use
     * @param modelPath  path to the model
     * @param llmPrefix  llm prefix to use
     */
    public LlmConfig(String modelTag, List<String> stopWords, String modelPath, String llmPrefix)
    {
        this(modelTag, stopWords, modelPath, llmPrefix, "", "", 4);
    }

    /**
     * Minimal constructor without numThreads
     *
     * @param modelTag   tag for the model
     * @param stopWords  stop words to use
     * @param modelPath  path to the model
     * @param llmPrefix  llm prefix to use
     * @param userTag    user tag to use
     * @param endTag     end tag to use
     */
    public LlmConfig(String modelTag, List<String> stopWords, String modelPath, String llmPrefix, String userTag, String endTag)
    {
      // Use 4 threads by default
      this(modelTag, stopWords, modelPath, llmPrefix, userTag, endTag, 4);
    }

    /**
     * Minimal constructor without userTag, and endTag
     *
     * @param modelTag   tag for the model
     * @param stopWords  stop words to use
     * @param modelPath  path to the model
     * @param llmPrefix  llm prefix to use
     * @param numThreads number of threads to use
     */
    public LlmConfig(String modelTag, List<String> stopWords, String modelPath, String llmPrefix, int numThreads)
    {
       this(modelTag, stopWords, modelPath, llmPrefix, "", "", numThreads);
    }

    /**
     * Main constructor
     *
     * @param modelTag   tag for the model
     * @param stopWords  stop words to use
     * @param modelPath  path to the model
     * @param llmPrefix  llm prefix to use
     * @param userTag    user tag to use
     * @param endTag     end tag to use
     * @param numThreads number of threads to use
     */
    public LlmConfig(String modelTag, List<String> stopWords, String modelPath,
                     String llmPrefix, String userTag, String endTag, int numThreads)
    {
          this.modelTag = modelTag;
          this.stopWords = stopWords;
          this.modelPath = modelPath;
          this.llmPrefix = llmPrefix;
          this.userTag = userTag;
          this.endTag = endTag;
          this.numThreads = numThreads;
    }

    /**
     * Gets the model tag.
     *
     * @return The model tag.
     */
    public String getModelTag()
    {
        return this.modelTag;
    }
    /**
     * Gets the user tag.
     *
     * @return The user tag.
     */
    public String getUserTag()
    {
        return this.userTag;
    }
    /**
     * Gets the end tag.
     *
     * @return The end tag.
     */
    public String getEndTag()
    {
        return this.endTag;
    }

    /**
     * Gets the list of stop words.
     *
     * @return The list of stop words.
     */
    public List<String> getStopWords()
    {
        return this.stopWords;
    }

    /**
     * Gets the model path.
     *
     * @return The model path.
     */
    public String getModelPath()
    {
        return this.modelPath;
    }

    /**
     * Gets the LLM prefix.
     *
     * @return The LLM prefix.
     */
    public String getLlmPrefix()
    {
        return this.llmPrefix;
    }

    /**
     * Gets the number of Threads used
     * @return The number of Threads LLM uses.
     */
    public int getNumThreads()
    {
        return this.numThreads;
    }

    /**
     * Sets the model tag.
     *
     * @param modelTag The model tag to set.
     */
    public void setModelTag(String modelTag)
    {
        this.modelTag = modelTag;
    }

     /**
     * Sets the user tag.
     *
     * @param userTag The user tag to set.
     */
    public void setUserTag(String userTag)
    {
        this.userTag = userTag;
    }

    /**
     * Sets the end tag.
     *
     * @param endTag The end tag to set.
     */
    public void setEndTag(String endTag)
    {
        this.endTag = endTag;
    }

    /**
     * Sets the list of stop words.
     *
     * @param stopWords The list of stop words to set.
     */
    public void setStopWords(List<String> stopWords)
    {
        this.stopWords = stopWords;
    }

    /**
     * Sets the model path.
     *
     * @param modelPath The model path to set.
     */
    public void setModelPath(String modelPath)
    {
        this.modelPath = modelPath;
    }

    /**
     * Sets the LLM prefix.
     *
     * @param llmPrefix The LLM prefix to set.
     */
    public void setLlmPrefix(String llmPrefix)
    {
        this.llmPrefix = llmPrefix;
    }

    /**
     * Sets the number of Threads.
     *
     * @param numThreads count of threads to use for LLM.
     */
    public void setNumThreads(int numThreads)
    {
        this.numThreads = numThreads;
    }
}

